### ISSUE TITLE
There is a small chance to mis-detect the position when a permanent write error happens

### ISSUE THREAD

**Issue by piste-jp-ibm** [LinearTapeFileSystem/ltfs#356]
**Describe the bug**

Tape drive is generally returns a permanent write error (WRITE PERM) as a deferred error. It means there are blocks on drive buffer but some of them cannot be written to tape drive.

LTFS tries to detect the last record on the tape when a WRITE PERM happens. And then cleanup (remove) extents which has unwritten blocks to the tape.

In the current code, LTFS fetches the number of record in the buffer and estimate the last record on the tape by the simple equation, `current position - number of records on the buffer`.

https://github.com/LinearTapeFileSystem/ltfs/blob/3b26f034e11a2a0cb767dfd914597b9a3527ffdc/src/iosched/unified.c#L2249-L2283

https://github.com/LinearTapeFileSystem/ltfs/blob/3b26f034e11a2a0cb767dfd914597b9a3527ffdc/src/tape_drivers/linux/sg/sg_tape.c#L4968

But this simple estimation might have a hole when a WRITE PERM happens at the begging of sequential write (first write or less than the size of the buffer). 

LTFS might remove extents even if it was written in the past when it happens. And it creates inaccessible contents from the latest index 

The scenario is 

1. LTFS writes some files (F11) and wrote an index successfully
2. LTFS reads a file except (F11) or unmount the tape
3. LTFS remounts the tape if tape is unmounted in the step2
4. LTFS writes a file (F12)
5. LTFS receives a WRITE PERM because of that a tape drive returns a WRITE PERM at the very beginning of the writing of records constructs F12
6. LTFS detects a record of F11 as the last record on the tape. Because drive might read tail records of F11 when drive moves to the append position implicitly

```
After Step 1:
|LABEL| INDEX #1 |F1 , F2 , F3 ... , F9 | INDEX #n-1 | <F10-(1)><F10-(2)>..<F10-(i)> , <F11-1><F11-2>.....<F11-(i-1)><F11-(i)> | INDEX #n |  

Before writing F12:
|..... , F9 | INDEX #n-1 | <F10-(1)><F10-(2)>..<F10-(i)> , <F11-1><F11-2>.....<F11-(i-1)><F11-(i)> | INDEX #n |
                                                                           ^----------------------------------^ 
Drive has records hi-lighted by `^---^` on buffer (buffer size is depends on the model of H/W)

After WRITE PERM writing F12:
|..... , F9 | INDEX #n-1 | <F10-(1)><F10-(2)>..<F10-(i)> , <F11-1><F11-2>.....<F11-(i-1)><F11-(i)> | INDEX #n |  <F12-1><F12-2>
                                                                                          ^----------------------------xxxxxxx^

`<F12-2>` cannot be written by the WRITE PERM but it is in the buffer. And the drive still has `<F11-(i)>` in the buffer. So LTFS detected the last record on the tape is `F11-(i)` even if it is already written successfully. And finally erase the extent that has `F11-(i)`.
```

As a result, similar behavior as #232 is made. 

**To Reproduce**

1. Write some files on the tape
2. Unmount the tape
3. Remount the tape
4. Write a file to the tape and cause an WRITE PERM at the fist physical write to the tape (from drive perspective)
5. Last written file on step1 goes to 0-byte

**Expected behavior**

All files written on the step1 shall be read correctly.

**Comment by piste-jp-ibm** [comment]
I found a good solution. I believe we can use the `LAST LOGICAL OBJECT LOCATION` field in the ` READ POSITION - 34h` command instead of the `NUMBER OF LOGICAL OBJECTS IN OBJECT BUFFER` field.

The spec stated like below
```
LAST LOGICAL OBJECT LOCATION:
After a write command, this field specifies the logical object address associated with the next logical object to be transferred from the buffer to the medium, in the range 0000 0000h to FFFF FFFFh. After a read command, this field specifies the logical object address associated with the last (most recent) data block or filemark to be transferred from the medium to the buffer. For any case where the buffer no longer contains a whole block of data or is empty, the value reported for the LAST LOGICAL OBJECT LOCATION is equal to the value reported for the FIRST LOGICAL OBJECT LOCATION.
```

Things to do in the fix are

1. Remove `*get_block_in_buffer()` I/F of the drive backend
2. Introduce `*get_next_block_to_transfer()` I/F
3. Capture the first record number that is not written on the tape after WRITE perm using the ``*get_next_block_to_transfer()` I/F` as `first_error_rec`
4. Evaluate every extents which has `first_error_rec` or larger record or not. And erase an extent if it has.
